### ISSUE TITLE
[BITV] 9.4.1.2/2.6 - The "Weather" menu - The implementation as an application menu is not complete as only role="menu" was used and also not appropriate in this case. Implementing it as a normal list with interactive elements would be a better alternative. (1)

### ISSUE THREAD

**Issue by AndyScherzinger** [nextcloud/server#37099]
<details>
<summary>Details</summary>

https://report.bitvtest.de/default-en/d63601ac-cb34-4645-8256-66bec78964a0.html#checkpoint-bfbf07e5c3-v2-n6

</details>

**Comment by JuliaKirschenheuter** [comment]
Dear @michaelnissenbaum,

I've looked through menu examples on the https://www.w3.org/WAI/tutorials/menus/ page and haven't seen any specific requirements for the action menu in our case, like: 

![Screenshot from 2023-10-30 14-14-33](https://github.com/nextcloud/server/assets/6078378/57b7c0f1-f775-49ab-a34b-6806e9251f75)
![Screenshot from 2023-10-30 14-11-36](https://github.com/nextcloud/server/assets/6078378/b5806099-653d-44ae-a1d5-f16dbe312214)

Would it be enough to leave a menu in form of the only list with "interactive elements"? Would it be enough to remove role="menu" on the `<ul>` elements to resolve this issue?

Thanks a lot!

**Comment by michaelnissenbaum** [comment]
Hello @JuliaKirschenheuter, In my opinion, if you remove the two roles "menu" and "menuitem," everything should be fine. Attached is the link to an article that describes the differences and use cases for both variants - https://inclusive-components.design/menus-menu-buttons/.

**Comment by JuliaKirschenheuter** [comment]
We have different types of menu:

1. Menu with action buttons, like:
![Screenshot from 2023-11-02 11-36-13](https://github.com/nextcloud/server/assets/6078378/c4258fe1-d21f-4314-b154-bd4bad7d331a)

2. Menu with mixed elements (action buttons,  input fields) like:
![Screenshot from 2023-11-02 11-41-48](https://github.com/nextcloud/server/assets/6078378/c87cf044-0cfa-4f4a-86be-13fa31f382fb)

3. Menu with links, like:
![Screenshot from 2023-11-02 11-45-13](https://github.com/nextcloud/server/assets/6078378/fd7c94db-bd2b-4881-a03d-dd33e39b90c3)
______

**Action menu and navigation menu functionalities shouldn't be mixed**
> https://marcus.io/blog/menu-or-not

If menu contains only links, then any `role="menu"` and `role="menuitem"` **should not be used.** 
> Don’t use ARIA, if the underlying HTML is sufficient for communicating what this thing is. In the case of a navigation menu, that’s a nested list of links. 
https://terrillthompson.com/1226
https://inclusive-components.design/menus-menu-buttons/

Requirements regarding menu button: https://www.w3.org/WAI/ARIA/apg/patterns/menu-button/

Does it mean, that `role="menuitem"` and `role="menu"` are needed only on nested navigation menus?
> https://www.w3.org/WAI/ARIA/apg/patterns/menubar/

**We should not use `role="menu"` and `role="menuitem"` on action menu with buttons and action menu with mixed components.**
> Also that role="menu" is a beast of its own, should be used with caution even in an "action menu" context, and that it does not always pay off to try something "smart" but to go for the simple route
https://marcus.io/blog/menu-or-not

> The issue with this pattern is that the document does not make it clear that this pattern should only be used when trying to replicate an application menu.
https://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html

Dear @michaelnissenbaum, just to be sure: have i understood correctly that we should remove `role="menu"` and `role="menuitem"` on every type of our menu (three examples above)? Thanks a lot!

**Comment by ShGKme** [comment]
We also have a classic menu like in text, which definitely should have `role="menu"`:
![image](https://github.com/nextcloud/server/assets/25978914/1fb49251-5c4e-4977-a601-5927de44a5bf)

**Comment by michaelnissenbaum** [comment]
Hi @JuliaKirschenheuter , the decision on which type of menu to use can be made based on the description of the "menu" role (ARIA 1.2) - "A menu is often a list of common actions or functions that the user can invoke. **The menu role is appropriate when a list of menu items is presented in a manner similar to a menu on a desktop application.**". If a menu is used for navigation purposes, then ARIA "menu" roles should not be used. The FlyOut Pattern is applied - https://www.w3.org/WAI/tutorials/menus/flyout/. My recommendation would be to use the "menu" role for Example 1 and not for 2 and 3.

**Comment by JuliaKirschenheuter** [comment]
Conclusion about which "menu" have to have `role="menu"` and `role="menuitem"`: 

1. Is not a menu and have to be re-done as usual NcSelect (follow-up ticket)
![image](https://github.com/nextcloud/server/assets/6078378/10cfae3d-1fb2-4103-a3a7-77ef676dee92)

2. Is not a menu because several components are mixed (input, action button...)
![image](https://github.com/nextcloud/server/assets/6078378/986d4172-4baa-4661-80a2-db0abee51cb0)

3. Not a menu because contains only links
![image](https://github.com/nextcloud/server/assets/6078378/fd7f058c-3c42-4fc1-a62f-848ef8816dfc)

4. Menu because have only action buttons inside
![image](https://github.com/nextcloud/server/assets/6078378/433ada62-4d74-41dd-89af-13dd31526272)

5. If there is a typical navigation -> should not be a menu too

______

Menu is only in case if all elements have only buttons. **It should looks like a "real desktop application menu"**

**Comment by ShGKme** [comment]
Some comments from my side.

`It should looks like a "real desktop application menu"` is not very specific, we have no menus like that. The only close to "desktop-like" is text with the menubar.

### Possible elements

In the actions we may have:
1. **Clickable items** that can have semantics of **menu items**
   - Links:
     - Hyperlink `NcActionLink`
     - In-app JS link `NcActionRouter`
   - Button `NcActionButton`
   - Menu form elements:
     - Checkbox: `NcActionCheckbox`
     - Radio: `NcActionRadio`
     - In Process: pressed state button (alternative to checkbox)
   - Additionally, a "radio group item". Can be considered a menu item with a sub-menu: `NcActionButtonGroup`
2. **Data input** items that should not be elements of `menu`:
   - Text, date, color, select: `NcActionInput`
   - Textarea: `NcActionTextEditable`
3. Misc.:
   - Text `NcActionText`
   - Separator `NcActionSeparator`
   - Caption `NcActionCaption`

  
### Criteria

Let's make such criteria to be a semantic menu.

`NcActions` is considered a `menu` if at the same time:
- It has **at least one `Clickable item`**
- It has **no `Data input`**
- It has **not only `Links`** as `Clickable items`, otherwise it is a navigation

Otherwise, it is just a popover.

### Examples

#### Menus

Menubar | Buttons | With checkbox and links | Buttons with text and separators | Button group
---|---|---|---|---
Text | File actions | Contact settings | Text | Mail
![image](https://github.com/nextcloud/server/assets/25978914/dff48cee-bcd9-4e29-9ab2-5c421d77e005) | ![image](https://github.com/nextcloud/server/assets/25978914/6e1cd6be-bc93-403c-89ce-83353699378c) | ![image](https://github.com/nextcloud/server/assets/25978914/89d90392-2ed6-4b55-af8c-55e46fe5a415) | ![image](https://github.com/nextcloud/server/assets/25978914/4f3a0484-ffa1-4977-9d10-cc80b711aabb) | ![image](https://github.com/nextcloud/server/assets/25978914/40b3191f-09f3-4962-aecd-101ba5ce279c)
. | user management | Mail | .
. | ![image](https://github.com/nextcloud/server/assets/25978914/0a7e370e-aa6a-4151-8c3a-42bdc0f468c1) | ![image](https://github.com/nextcloud/server/assets/25978914/1dc77cf6-dbcd-4df6-80e4-e0d3d140af48) | . | .

#### Navigation (no role)

![image](https://github.com/nextcloud/server/assets/25978914/14f984db-e091-4a55-a062-eba84235e47e)

#### Just popovers (no role)

Weather status | Group edit
---|---
![image](https://github.com/nextcloud/server/assets/25978914/c9c4e390-e86f-40b3-abfc-bb04ef8c10e2) | ![image](https://github.com/nextcloud/server/assets/25978914/e29a92db-b2cf-4237-b8da-fad0b14bda97)
